### PR TITLE
add fast uint256_add function

### PR DESCRIPTION
**Pull request by feltroidprime**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo-lang/139)

### CODE REVIEW

**Review by spapinistarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo-lang/139)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @feltroidprime)

___
*[`src/starkware/cairo/common/uint256.cairo` line 68 at r1](https://reviewable.io/reviews/starkware-libs/cairo-lang/139#-NLugySx9v_fK1E6uQyg:-NLugySx9v_fK1E6uQyh:b-1ksk1) ([raw file](https://github.com/starkware-libs/cairo-lang/blob/e28dc7db1b5862be658e7d16368f1de7ee8a62f1/src/starkware/cairo/common/uint256.cairo#L68)):*
> ```Cairo
>     if (has_carry_low != 0) {
>         if (has_carry_high != 0) {
>             assert res.low = res_low - SHIFT;
> ```

Remove the assert keywords, so the compilet will make sure it's 1 instruction.
___
*[`src/starkware/cairo/common/uint256.cairo` line 68 at r1](https://reviewable.io/reviews/starkware-libs/cairo-lang/139#-NLuhBBb3dNLzrpH1Lwn:-NLuhBBb3dNLzrpH1Lwo:b56snke) ([raw file](https://github.com/starkware-libs/cairo-lang/blob/e28dc7db1b5862be658e7d16368f1de7ee8a62f1/src/starkware/cairo/common/uint256.cairo#L68)):*
> ```Cairo
>     if (has_carry_low != 0) {
>         if (has_carry_high != 0) {
>             assert res.low = res_low - SHIFT;
> ```

This can be optinized a bit by making sure res is located correctly on the stack.
For example, do tempvar range_check_ptr=range_check_ptr+2 before all the ifs.
Then, instead of saving res as a local, save it as a tempvar in the ifs:
Tempvar low = bla
Tempvar high = bla
Let res =Uint256(low, high)

**Review by ilyalesokhin-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo-lang/139)* status: 0 of 1 files reviewed, 4 unresolved discussions (waiting on @feltroidprime and @spapinistarkware)

___
*[`src/starkware/cairo/common/uint256.cairo` line 53 at r1](https://reviewable.io/reviews/starkware-libs/cairo-lang/139#-NLuWEdX4H-wfkUXNpLt:-NLuWEdX4H-wfkUXNpLu:btxxxws) ([raw file](https://github.com/starkware-libs/cairo-lang/blob/e28dc7db1b5862be658e7d16368f1de7ee8a62f1/src/starkware/cairo/common/uint256.cairo#L53)):*
> ```Cairo
> }
> // Adds two integers. Returns the result as a 256-bit integer and the (1-bit) carry.
> func uint256_fast_add{range_check_ptr}(a: Uint256, b: Uint256) -> (res: Uint256, carry: felt) {
> ```

Mention that this is a version of uint256_add that is optimized for speed rather than for size.



_Code quote:_
```Cairo
 uint256_fast_add
```
___
*[`src/starkware/cairo/common/uint256.cairo` line 100 at r1](https://reviewable.io/reviews/starkware-libs/cairo-lang/139#-NLu_OX43DzuTG3cqQCV:-NLu_OX43DzuTG3cqQCW:bhbcipy) ([raw file](https://github.com/starkware-libs/cairo-lang/blob/e28dc7db1b5862be658e7d16368f1de7ee8a62f1/src/starkware/cairo/common/uint256.cairo#L100)):*
> ```Cairo
>     }
> 
>     return (res, carry_high);
> ```

Mistake?



_Code quote:_
```Cairo
return (res, carry_high);
```

File: src/starkware/cairo/common/uint256.cairo [hunk 1]
```diff
@@ -49,6 +49,56 @@ func uint256_add{range_check_ptr}(a: Uint256, b: Uint256) -> (res: Uint256, carr
...
+            assert res.high = res_high;
+            assert [range_check_ptr] = res.low;
+            assert [range_check_ptr + 1] = res.high;
+            let range_check_ptr = range_check_ptr + 2;
+            return (res, 0);
+        }
+    }
+
+    return (res, carry_high);
+}
```
Comment (Oppen) [inline src/starkware/cairo/common/uint256.cairo:101]:
```suggestion
    if (has_carry_low != 0) {
        assert res.low = res_low - SHIFT;
        assert res.high = res_high + 1 - SHIFT;
        assert [range_check_ptr] = res.low;
        assert [range_check_ptr + 1] = res.high;
        let range_check_ptr = range_check_ptr + 2;
        if (has_carry_high != 0) {
            return (res, 1);
        } else {
            return (res, 0);
        }
    } else {
        assert res.low = res_low;
        assert res.high = res_high - SHIFT;
        assert [range_check_ptr] = res.low;
        assert [range_check_ptr + 1] = res.high;
        let range_check_ptr = range_check_ptr + 2;
        if (has_carry_high != 0) {
            return (res, 1);
        } else {
            return (res, 0);
        }
    }

    return (res, carry_high);
}
```

**Review by ilyalesokhin-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo-lang/139)* status: 0 of 1 files reviewed, 5 unresolved discussions (waiting on @feltroidprime, @Oppen, and @spapinistarkware)

___
*[`src/starkware/cairo/common/uint256.cairo` line 68 at r1](https://reviewable.io/reviews/starkware-libs/cairo-lang/139#-NLuhBBb3dNLzrpH1Lwn:-NLujh_dEYaomlEgiEND:b-9v4xzi) ([raw file](https://github.com/starkware-libs/cairo-lang/blob/e28dc7db1b5862be658e7d16368f1de7ee8a62f1/src/starkware/cairo/common/uint256.cairo#L68)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

This can be optinized a bit by making sure res is located correctly on the stack.
For example, do tempvar range_check_ptr=range_check_ptr+2 before all the ifs.
Then, instead of saving res as a local, save it as a tempvar in the ifs:
Tempvar low = bla
Tempvar high = bla
Let res =Uint256(low, high)
</blockquote></details>

consider:
func uint256_fast_add{range_check_ptr}(a: Uint256, b: Uint256) -> (res: Uint256, carry: felt) {
    alloc_locals;
    local has_carry_low: felt;
    local has_carry_high: felt;
    let res_low = a.low + b.low;
    let res_high = a.high + b.high;
    %{
        has_carry_low = 1 if ids.res_low >= ids.SHIFT else 0
        ids.has_carry_low=has_carry_low
        ids.has_carry_high = 1 if (ids.res_high+has_carry_low) >= ids.SHIFT else 0
    %}

    // In the following code we put range_check_ptr we put range_check_ptr in the
    // right place on the stack for return value optimization.
   
    if (has_carry_low != 0) {
            tempvar res_low = res_low;
            tempvar res_high = res_high;
        if (has_carry_high != 0) {
            tempvar range_check_ptr = range_check_ptr + 2;
            tempvar res = Uint256(low=res_low - SHIFT, high=res_high + 1 - SHIFT);
            assert [range_check_ptr - 2] = res.low;
            assert [range_check_ptr - 1] = res.high;
            return (res, 1);
        } else {
            tempvar range_check_ptr = range_check_ptr + 2;
            tempvar res = Uint256(low=res_low - SHIFT, high=res_high + 1 );
            assert [range_check_ptr - 2] = res.low;
            assert [range_check_ptr - 1] = res.high;
            return (res, 0);
        }
    } else {
        if (has_carry_high != 0) {
            tempvar res_high = res_high;
            tempvar range_check_ptr = range_check_ptr + 2;
            tempvar res = Uint256(low=res_low, high=res_high - SHIFT);
            assert [range_check_ptr - 2] = res.low;
            assert [range_check_ptr - 1] = res.high;
            return (res, 1);
        } else {
            tempvar range_check_ptr = range_check_ptr + 2;
            tempvar res = Uint256(low=res_low, high=res_high);
            assert [range_check_ptr - 2] = res.low;
            assert [range_check_ptr - 1] = res.high;
            return (res, 0);
        }
    }
}

File: src/starkware/cairo/common/uint256.cairo (hunk 1 above)
Comment (feltroidprime) [reply src/starkware/cairo/common/uint256.cairo:101]: line     return (res, carry_high); is indeed a copy paste mistake

**Pull request closed without merging**